packages/imperative/src/logger/src/log4jsToWinston.ts

```diff
@@ -0,0 +1,273 @@
+        ); // Timezone offset
+    };
+
+    return format.combine(
+        format.timestamp(), // Add timestamp to the info object, default format (ISO)
+        format.printf((info) => {
+            // Process the pattern for each log entry
+            let output = pattern;
+
+            // Handle colorization tokens first (%[ and %]) - simple removal
+            output = output.replace(/%\[/g, "").replace(/%\]/g, "");
+
+            const dateTokenRegex = /%d(?:\{([^}]+)\})?/g; // Regex specifically for date tokens
+            output = output.replace(
+                dateTokenRegex,
+                (dateToken, capturedFormat): string => {
+                    // Default format for plain %d
+                    let formatString = "YYYY-MM-DDTHH:mm:ss.SSS";
+                    if (capturedFormat) {
+                        // If a format like %d{...} is found, map it
+                        formatString = mapLog4jsToDayjsFormat(capturedFormat);
+                    }
+                    // Use dayjs to format the timestamp with the specific format for this token
+                    return dayjs(info.timestamp as string).format(
+                        formatString
+                    );
+                }
+            );
+
+            // Regex for non-date, non-color tokens handled previously
+            const otherTokenRegex = /(%p|%c|%m|%n|%h|%z|%%)/g;
+            output = output.replace(otherTokenRegex, (token): string => {
+                // Handle remaining tokens
+                switch (token) {
+                    case "%p":
+                        return String(info.level).toLocaleUpperCase();
+                    case "%m":
+                        return String(info.message);
+                    case "%n":
+                        return "\n";
+                    case "%c":
+                        return String(info.category ?? "default");
+                    case "%h":
+                        return os.hostname();
+                    case "%z":
+                        return String(process.pid);
+                    case "%%":
+                        return "%";
+                }
+            });
+
+            // Add category to info object if not present and %c is in the original pattern
+            if (!info.category && pattern.includes("%c")) {
+                info.category = "default";
+            }
+            return output;
+        })
+    );
+}
+
+/**
+ * Type guard for log4js appender with a pattern layout.
+ */
+export function hasPatternLayout(
+    appender: unknown
+): appender is { layout: { type: "pattern"; pattern: string } } {
+    return (
+        typeof appender === "object" &&
+        appender !== null &&
+        "layout" in appender &&
+        typeof (appender as any).layout === "object" &&
+        (appender as any).layout !== null &&
+        (appender as any).layout.type === "pattern" &&
+        typeof (appender as any).layout.pattern === "string"
+    );
```

Comment:
I like the use of type guards here. Should we define it as a type?
i.e. `const type tAppenderPatter = { layout: { type: "pattern"; pattern: string } }`

Also, couldn't we simplify the function using optional-chaining (Elvis operator?)?

```typescript
const layout = appender?.layout;
if (layout != null) {
  return layout.type === "pattern" && typeof layout.pattern === "string" 
}
return false;
```

Reply:
This won't work because `type` and `pattern` are not properties of `object`.
That said, I did clean up the logic in an upcoming commit and added a separate type guard for `layout is PatternLayout` (new type) for readability 😋